packages/vitest/src/node/mocker.ts

```diff
@@ -53,29 +56,57 @@ function mockPrototype(proto: any) {
   return newProto
 }
 
-function mockObject(obj: any) {
-  const type = getObjectType(obj)
+function mockObject(obj: any, seen = new WeakSet()) {
+  if (seen.has(obj))
+    return obj
 
   if (Array.isArray(obj))
     return []
-  else if (type !== 'Object' && type !== 'Module')
+
+  const objectType = getObjectType(obj)
+
+  if (objectType !== 'Object' && objectType !== 'Module' && objectType !== 'Function')
     return obj
 
+  seen.add(obj)
+
   const newObj = { ...obj }
 
   const proto = mockPrototype(Object.getPrototypeOf(obj))
   Object.setPrototypeOf(newObj, proto)
 
   // eslint-disable-next-line no-restricted-syntax
-  for (const k in obj) {
-    newObj[k] = mockObject(obj[k])
-    const type = getObjectType(obj[k])
-
-    if (type.includes('Function') && !obj[k].__isSpy) {
-      spyOn(newObj, k).mockImplementation(() => {})
-      Object.defineProperty(newObj[k], 'length', { value: 0 }) // tinyspy retains length, but jest doesnt
+  for (const objAttr in obj) {
+    if (!Object.prototype.hasOwnProperty.call(obj, objAttr)) continue
+
+    const objValue = obj[objAttr]
+    const nestedObjectType = getObjectType(objValue)
+
+    if (nestedObjectType === 'Function') {
+      // ex: function foo() {}
+      // foo.newFunction = () => {}
+      const hasFunctionsAttr = Object.keys(objValue).length !== 0
+
+      if (hasFunctionsAttr) {
+        const internalFunctionsMocked = mockObject(objValue, seen)
+        const fn = newObj[objAttr]
+        // eslint-disable-next-line no-restricted-syntax
+        for (const internalFnKey in internalFunctionsMocked) {
+          if (Object.prototype.hasOwnProperty.call(internalFunctionsMocked, internalFnKey))
+            fn[internalFnKey] = internalFunctionsMocked[internalFnKey]
+        }
+        newObj[objAttr] = mockObject(objValue, seen)
```

Comment:
not sure if calling `axios` will work then? since it doesnt spy on the call itself, but only on attributes?

Reply:
@sheremet-va You're right. I will check it again. Nice catch!